### PR TITLE
Refactor database interface to prefer blockchain.HeadFetcher

### PR DESCRIPTION
**Pull request by prestonvanloon**
This PR is an effort to use a single source of truth for "head" information.

Key changes:
- Refactor Database interface into limiting access composable interfaces. 
- Change all production references of `db.Database` to the appropriate db access.
- Change all usage of HeadState / HeadBlock / HeadRoot to use the HeadFetcher, especially in RPC services where it is most critical.
- Limit access to database Head to only HeadFetcher chain service and powchain service. The powchain service needs it because it is a circular dependency if powchain uses the chain service. This could be a future refactoring opportunity.
- Delete CanonicalRoot method from chain service. It was unused.

Things to look out for at runtime:
- If the beacon chain is complaining about signatures not verifying, this is a discrepancy in expected duties. 
- Panic / nil pointer issues.

### CODE REVIEW

File: beacon-chain/db/iface/interface.go [hunk 1]
```diff
@@ -13,82 +13,112 @@ import (
...
+	JustifiedCheckpoint(ctx context.Context) (*eth.Checkpoint, error)
+	FinalizedCheckpoint(ctx context.Context) (*eth.Checkpoint, error)
+	// Archival data handlers for storing/retrieving historical beacon node information.
+	ArchivedActiveValidatorChanges(ctx context.Context, epoch uint64) (*ethereum_beacon_p2p_v1.ArchivedActiveSetChanges, error)
+	ArchivedCommitteeInfo(ctx context.Context, epoch uint64) (*ethereum_beacon_p2p_v1.ArchivedCommitteeInfo, error)
+	ArchivedBalances(ctx context.Context, epoch uint64) ([]uint64, error)
+	ArchivedValidatorParticipation(ctx context.Context, epoch uint64) (*eth.ValidatorParticipation, error)
+	// Deposit contract related handlers.
+	DepositContractAddress(ctx context.Context) ([]byte, error)
+	//Powchain operations
```
Comment (terencechain) [inline beacon-chain/db/iface/interface.go]:
```suggestion
	// Powchain operations.
```

File: beacon-chain/db/iface/interface.go [hunk 2]
```diff
@@ -13,82 +13,112 @@ import (
...
-	ArchivedCommitteeInfo(ctx context.Context, epoch uint64) (*ethereum_beacon_p2p_v1.ArchivedCommitteeInfo, error)
 	SaveArchivedCommitteeInfo(ctx context.Context, epoch uint64, info *ethereum_beacon_p2p_v1.ArchivedCommitteeInfo) error
-	ArchivedBalances(ctx context.Context, epoch uint64) ([]uint64, error)
 	SaveArchivedBalances(ctx context.Context, epoch uint64, balances []uint64) error
-	ArchivedValidatorParticipation(ctx context.Context, epoch uint64) (*eth.ValidatorParticipation, error)
 	SaveArchivedValidatorParticipation(ctx context.Context, epoch uint64, part *eth.ValidatorParticipation) error
 	// Deposit contract related handlers.
-	DepositContractAddress(ctx context.Context) ([]byte, error)
 	SaveDepositContractAddress(ctx context.Context, addr common.Address) error
 	//Powchain operations
```
Comment (terencechain) [inline beacon-chain/db/iface/interface.go]:
```suggestion
	// Powchain operations.
```

**Comment by prestonvanloon** [comment]
One thing I need to add still is that saveHead needs to make a copy of the head state to put in cache, otherwise it could still be mutated somehow. Same thing with head block

**Comment by prestonvanloon** [comment]
Regarding saveHead, we only need to clone the inputs. In this method, only the signed head block is an input where the head state was fetched from the DB. Only the block needs to be cloned so this should be good to go now.

**Pull request merged**